### ISSUE TITLE
Thread safe sockets

### ISSUE THREAD

**Issue by alexhultman** [uWebSockets/uWebSockets#135]
Currently all functions of Socket are thread unsafe.

**Comment by joyousrabbit** [comment]
following

**Comment by alexhultman** [comment]
I've looked at this a bit more and it seems more complex than I first thought. It's hard to do efficiently and it would be better for you to be able to control how you queue things and how you copy things.

Instead of having thread safe functions (which would need one mutex per socket), I would want to just add helper functions to the Server to allow easier use of the uv_async_t thing. If you optimize things correctly you should be able to only have one mutex instead of one per socket, it depends on what you are doing and in worst case you will have to add your own mutex via the Socket.setData.

I've done similar things in libwebsockets, that library is not thread safe or safe in any way really, so this is how you would to it:

* You need one queue where you store pairs of sockets and message
* You need one mutex for the queue

// register async handler (this function is always called from the correct thread)
server.onAsync([] {
// loop over all queued pairs, send the message using the socket
});

// when you need to send from a separate thread:
* Lock the mutex, queue the message and the socket
* Call server.async();

Calling server.async() will trigger the function registered with onAsync from the thread that belongs to the server.

It sounds complex, yes, but it allows a lot more optimizations and control and you could tailor it to your needs.

**Comment by alexhultman** [comment]
This would be solution number 1, later I could look at this again because it can be optimized a lot more but this will require a good look at the sending of everything.

**Comment by alexhultman** [comment]
Scratch that, this problem is really two things: socket / socket thread safety and socket / server thread safety.

* With socket / socket thread safety I mean "can I send to the same socket on multiple threads?". Answer to this is: no, never, will not be supported. Use your own mutex if you need this, attach it using socket.setData.

* With socket / server thread safety I mean "can this socket send from another thread, while the server runs on another thread". Answer here is: no, but I should be able to fix this up pretty nicely.

So yeah, it looks bright. All you would need if you want "full" thread safety is to attach your own mutex and lock it before you send and unlock it when you have sent.

**Comment by alexhultman** [comment]
The only thing that is really non-safe is in the write function where I call uv_poll_start when the kernel buffer is full. This call would need to be moved inside an uv_async_t and called from the servers thread. This should happen when the server is "non-master", but not when being master so there should be minimal if any performance loss actually.

**Comment by alexhultman** [comment]
Nope, there needs to be a mutex in every socket. So yeah, it is tricky - either you lose memory or you lose performance.

**Comment by alexhultman** [comment]
Derp, I need to think about this for a while :wink:

**Comment by joyousrabbit** [comment]
And a webSocket server just has hundreds or thousands of socket
connections, a polling will not hurt its performance. I once worked in a
network game company I know they do like this.

2016-06-02 9:26 GMT+02:00 Jun FU <fujunde@gmail.com>:

**Comment by joyousrabbit** [comment]
Hello Alex,

     I have another question. It seems server.send() works in blocking
mode. So when broadcasting, even only one client is blocked, all the queue
from the server side will also be blocked, right?

     Can we send message at noblocking mode?

     Best regards,
     Jun

2016-06-02 9:29 GMT+02:00 Jun FU <fujunde@gmail.com>:

**Comment by alexhultman** [comment]
Send is nonblocking. You cannot send from another thread

**Comment by AlexVK79** [comment]
> Send is nonblocking. You cannot send from another thread

If we can't send from another thread how do we organise a C++ game loop on server?

`server.run();`
will block the main thread, so while we do have async handlers like onMessage they are best suited for getting input from client, but we need to run game logics/physics at some fixed time intervals.

I would assume that we run a second threas like this:
`void game_thread_func()
{
while (1)
{
    update_physics();
    server.broadcast(worldState,...);
}
}`

But if we can't send/broadcast from another thread, what's the best way to do this ?
I guess we need some variant of Server::run() that is non-blocking and will run
in our own while() {} loop in the main threadwhere we also do game logics and broadcast?

**Comment by alexhultman** [comment]
You would register a timer to the underlying uv loop. Or, if you need to send from another thread you use uv_async_t. You need to write some libuv code to get these things working. If you run things on the main thread you can access the libuv loop via uv_default_loop().

**Comment by AlexVK79** [comment]
Thanks! Timer approach seems to work good so far.

**Comment by alexhultman** [comment]
The problem that needs to be solved is, WebSocket::send modifies itself (it is not thread safe), **but it also modifies the server** if the message does not fit in the kernel buffer (it needs to poll for writable).

* I do not want to make every socket thread safe (this would be a massive overhead and a lot of waste in most cases. The user knows how their data behaves and they can use mutexes appropriately to solve this).

* I do want to solve the problem of modifying **the server**. Calling WebSocket::send should be possible from any thread without messing up the server of another thread, it should only mess up itself (the WebSocket), which could be controlled with mutexes.

This adds a small overhead, but I still want to keep that overhead as opt-in via a server option.

@joyousrabbit In what way are you using threads and what problem are you facing?

**Comment by joyousrabbit** [comment]
@alexhultman 
Thanks for asking. I solved the problem by seril the recv and send.

I have car server, it receives cmd from client-end, and also send car status to client. There is a variable used accessed by both direction ( the client's status). So conflict sometimes between recv and send.

Then I solved this by add a common lock for recv and send. When recv, it will not send. When send, it will not recv. Do you think that this is a good solution?

**Comment by AlexVK79** [comment]
Many cloud instances are single core, so having  the best performance on them is way better than thread safety out of the box. That said, a server option to enable it would not hurt unless it affects performance (of the unsafe variant).

**Comment by alexhultman** [comment]
@joyousrabbit It might work. As long as the loop is frozen inside a callback handler when you modify it from other threads. You can only modify the server when it is inside a callback handler, so if you lock and block that handler while you send from another thread, that will work.

So if you get a message, block the handler, send from another thread and unblocks the handler it will equal "sending from the same thread". But I will still take a look into this to make it easier to work with.

**Comment by alexhultman** [comment]
Basically uv_poll_start might be called in WebSocket::send and this function can only be called when the owning server is inside a callback handler.

**Comment by socmag** [comment]
It's impressive in my server that a different thread writes to the socket than the receiving thread,  but I can  guarantee that only one thread will issue sends for the same socket. 

From what I am reading here,  this usr case is (or will be)  supported right?

**Comment by socmag** [comment]
Imperative  not impressive,  although uWS is impressive :P

**Comment by alexhultman** [comment]
The plan is to support sending from different threads yes. Currently it is something you will need to manage yourself via uv_async_t but I know how to solve this efficiently.

**Comment by socmag** [comment]
awesome, cheers!

**Comment by socmag** [comment]
We got it working as you suggested,  thanks. 

One question that arose is how to keep a reference to the WebSocket after the callback has returned in order that we can write to it asynchronously at a future point. 

Right now we have to make a change to the sources to allow us to make the uv_poll_t and the constructor public so we can steal it and pass it around threads .  We go to a lot of trouble to make sure that pointer is still usable down the line. 

That is fine,  and works pretty  well,  however we want you to know that we do this so that you can decide if making those available in the core is the right thing to do,  or something else. 

Additionally,  though unrelated,  it  might be good  if at some point in the future there was a way to get uWS to  send directly from application buffers to avoid that malioc in  the write function. 

I realize this is a pretty advanced use case but we can control the lifetime of our buffers and avoid allocations and copies  of we do that. 

Low priority but just a thought. 

Is working quite well though :-)

**Comment by socmag** [comment]
By the way,  setData/get data is much appreciated.  Thanks for thinking of that.  It allows us to avoid a really nasty mutex.

**Comment by alexhultman** [comment]
What's wrong with passing the WebSocket around?

By attaching your own data structure at connection event via setData, and marking that data somehow at disconnection event you can easily track the status of your sockets. Imagine having a shared_ptr attached and then passed around - that way you get an easy tracking of the status of the socket.

**Comment by socmag** [comment]
I don't have the code in front of me right now, but yes, that is a sane
idea; however we get compiler errors in the lambda at the moment due to
const correctness checks.

Maybe you could take a look at your callback std::function definition,
maybe you have the WebSocket argument defined as const?

BTW,  probably you should support move semantics on that thing.

**Comment by socmag** [comment]
With regard to the second suggestion,  that's a good idea.  We do something
similar already,  manually ref counting our application client state.

The issue is set/getData is just a void pointer at the moment.  I don't
think that will play well with shared_ptr?

**Comment by socmag** [comment]
Maybe it would be good to be able to templatize WebSocket on
application_state_t?

**Comment by alexhultman** [comment]
The memory management is up to the user. It's just like libwebsockets, you need to track everything manually. I solve it by attaching my own user data at connection which holds the WebSocket and in the case of non-GC languages, also holds some kind of reference or some kind of boolean to signal if the socket is valid or not:

struct UserData {
WebSocket webSocket;
int ref = 1;
};

onConnection:
webSocket.setData(new UserData({webSocket}));

onDisconnection:
((UserData *) webSocket.getData()).ref--;

Then you pass the UserData around and use the webSocket it contains if the ref is still valid.

The above is just a very simple example and can probably be solved in a better way. The main point is, in C++ you have no GC so there is no place to store if a websocket is valid or not, and if you store that info you need to delete it yourself.

How would a socket know if it has been deleted? If the socket has been deleted it cannot hold any info to singal it has been deleted. The only way to solve this is to manage this tracking yourself. You can use shared_ptr, you just have to call the destructor yourself since it's stored as a void *. shared_ptr is simple but also not really that performant.

**Comment by alexhultman** [comment]
Omg, the above is kind of messed up but you get the idea. You need to track the validity of the webSocket yourself.

**Comment by socmag** [comment]
Yes,  that philosophy is totally acceptable and pretty much* exactly* what
we are doing.

Couple of issues :

One apparently atm we can't call "send" in a lambda where the WebSocket is
captured in OnConnection.. I'll look in to that tomorrow.

Second, you say memory management is up to the user,  but the malloc for
large buffers in the send function says otherwise.  Probably we should be
using "write" directly instead of send? Will look into that as well. We
don't mind calling format buffer manually, and ensuring the layout etc.

Totally agree with your assumption that it is better to manage lifetime
yourself than rely on shared_pointer. It's working fine the way it is,
fast, and the semantics are clear.

Still, not a bad idea to templatize on the user data type and default to
void*. Even if we don't use it,  could be nice for some people.

Okay, time for bed here.  Will play tomorrow.

Thanks Alex!

**Comment by alexhultman** [comment]
Aha, you mean like Server<UserData> s; and onConnection([](WebSocket<SERVER, UserData>) webSocket)? That is a little verbose but could probably be done yes, then you can store whatever type you want. I like that.

The send could be improved to allow passing memory the user owns, I haven't looked into that but yes, the write and formatMessage functions are a good start but we should probably wrap that up into a public function or something. The interface is very much flexible and I accept PR's. I haven't finished the client support merge yet, but as soon as that is finished I can start doing a lot of different stuff again. TBH I haven't done anything lately.

**Comment by jimmyt857** [comment]
I'm not against this idea, but just want to add that I think you would need to move most of the code into header files if you do that, so the compiler can build a version of the server/websocket classes for the user's data type when they compile it.

**Comment by jimmyt857** [comment]
As for thread safety, I have a utility function that I wrote for my own purposes:

```
// these would go in the EventSystem
std::vector<std::function<void()>> mainThreadTasks;
std::mutex mainThreadLock;
uv_async_t async;

// call this in EventSystem constructor
uv_async_init(loop, &async, RunMainThreadTasks);

    // run all of the user's tasks
    void RunMainThreadTasks(uv_async_t *handle)
    {
        mainThreadLock.lock();
        for (const auto &task : mainThreadTasks)
        {
            try {
                task();
            } catch (std::runtime_error e) {
               std::cout << e.what() << std::endl;
            }
        }
        mainThreadTasks = std::vector<std::function<void()>>();
        mainThreadLock.unlock();
    }

    // user calls this to do their work from another thread
    void RunOnMainThread(std::function<void()> callback)
    {
        mainThreadLock.lock();
        mainThreadTasks.push_back(callback);
        mainThreadLock.unlock();
        uv_async_send(&async);
    }
```

This would add a little bit of memory to the EventSystem object, and allow the user to do whatever they want by running their work on the same event loop. It might bother some users to move a lot of their work to callbacks, however.

**Comment by alexhultman** [comment]
@jimmyt857 Yep, that's pretty much exactly what you have to do. The uv_poll_start call in WebSocket::write needs to be wrapped in RunOnMainThread([]() {
// do the uv_poll_start here
});

And the mutex, vector and uv_async_t is then per *Server*.

**Comment by alexhultman** [comment]
And to minimize the usage of uv_async_t one can just compare pthread_self with the pthread_self of the thread that runs the loop of the server which the websocket originates from. That's like 2 mov instructions according to StackOverflow.

**Comment by alexhultman** [comment]
With this commit you should be able to send from any thread, if you pass true as second argument to EventSystem constructor: 2d66230102b86260df92acfa0c57de11fe853fb4

**Comment by alexhultman** [comment]
I need to clean up the concept of what is safe and what is unsafe though. It's a mess right now.

**Comment by socmag** [comment]
okay great, i'll check it out right now

**Comment by socmag** [comment]
works like a charm, thanks @alexhultman!

**Comment by alexhultman** [comment]
So you still need to attach your own mutex per socket if you need "race free" sends, but now it should at least not mess up from different threads.

**Comment by socmag** [comment]
Of course, that is totally fine Alex.  Thanks.

**Comment by agauniyal** [comment]
Sorry for bumping up this thread, is the second `true` param still needed? Because I don't see it mentioned in multithreaded example `EventSystem es(MASTER);`.

**Comment by alexhultman** [comment]
Basically you specify WORKER for somewhat thead safe server and MASTER for single threaded servers. WORKER has thread safe upgrade and close and its sockets can send from different threads but still get race condition if sending at the same time (you need to attach your own mutex to each socket with setData for that). Best perf is to keep a group of sockets on the same thread, so if you have a server with 4 threads you might want to keep all sockets that belong to one thread separated from the other threads. It's a mess, really.

**Comment by agauniyal** [comment]
Ah, okay!
